src/jdk.compiler/share/classes/com/sun/tools/javac/comp/AttrRecover.java

```diff
@@ -0,0 +1,276 @@
+import com.sun.tools.javac.code.TypeTag;
+import com.sun.tools.javac.code.Types;
+import com.sun.tools.javac.comp.Attr.ResultInfo;
+import com.sun.tools.javac.comp.DeferredAttr.AttrMode;
+import com.sun.tools.javac.tree.JCTree;
+import com.sun.tools.javac.tree.JCTree.JCBlock;
+import com.sun.tools.javac.tree.JCTree.JCClassDecl;
+import com.sun.tools.javac.tree.JCTree.JCErroneous;
+import com.sun.tools.javac.tree.JCTree.JCExpression;
+import com.sun.tools.javac.tree.JCTree.JCLambda;
+import com.sun.tools.javac.tree.JCTree.JCMethodInvocation;
+import com.sun.tools.javac.tree.JCTree.JCReturn;
+import com.sun.tools.javac.tree.JCTree.JCVariableDecl;
+import com.sun.tools.javac.tree.JCTree.Tag;
+import com.sun.tools.javac.tree.TreeInfo;
+import com.sun.tools.javac.tree.TreeMaker;
+import com.sun.tools.javac.tree.TreeTranslator;
+import com.sun.tools.javac.util.Context;
+import com.sun.tools.javac.util.JCDiagnostic;
+import com.sun.tools.javac.util.List;
+import com.sun.tools.javac.util.ListBuffer;
+import com.sun.tools.javac.util.Names;
+
+/** This is an error recovery addon for Attr. Currently, it recovers
+ *  method invocations with lambdas, that require type inference.
+ *
+ *  <p><b>This is NOT part of any supported API.
+ *  If you write code that depends on this, you do so at your own risk.
+ *  This code and its internal interfaces are subject to change or
+ *  deletion without notice.</b>
+ */
+public class AttrRecover {
+    protected static final Context.Key<AttrRecover> attrRepairKey = new Context.Key<>();
+
+    final Attr attr;
+    final DeferredAttr deferredAttr;
+    final Names names;
+    final TreeMaker make;
+    final Symtab syms;
+    final Types types;
+
+    public static AttrRecover instance(Context context) {
+        AttrRecover instance = context.get(attrRepairKey);
+        if (instance == null)
+            instance = new AttrRecover(context);
+        return instance;
+    }
+
+    protected AttrRecover(Context context) {
+        context.put(attrRepairKey, this);
+
+        attr = Attr.instance(context);
+        deferredAttr = DeferredAttr.instance(context);
+        names = Names.instance(context);
+        make = TreeMaker.instance(context);
+        syms = Symtab.instance(context);
+        types = Types.instance(context);
+    }
+
+    private final ListBuffer<RecoverTodo> recoveryTodo = new ListBuffer<>();
+
+    public void doRecovery() {
+        while (recoveryTodo.nonEmpty()) {
+            RecoverTodo todo = recoveryTodo.remove();
+            ListBuffer<Runnable> rollback = new ListBuffer<>();
+            boolean repaired = false;
+            RECOVER: if (todo.env.tree.hasTag(Tag.APPLY)) {
+                JCMethodInvocation mit = (JCMethodInvocation) todo.env.tree;
+                if ((todo.candSym.flags() & Flags.VARARGS) == 0 &&
+                    mit.args.length() > todo.candSym.type.getParameterTypes().length()) {
+                    break RECOVER; //too many actual parameters, skip
```

Comment:
I like the patch after the last iteration. My only comment would be about test: TestGetTypeMirrorReference.java which does a golden file like comparison but I understand that there are not many options. I would suggest though to place this test along with file: TestGetTypeMirrorReferenceData.java in a separate folder and add some doc to the test so that it would be easier to modify it in the future

Reply:
Thanks!

I've tried to improve the text (moved to a separate directory, added comment) in 988f85b.